### ISSUE TITLE
Weird error with piSort and generalization

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#3940]
```agda
open import Data.List.Base                                  using (List; []; _∷_)
open import Data.List.All                                   using (All; []; _∷_)
open import Data.List.Relation.Binary.Sublist.Propositional using (_⊆_; []; _∷ʳ_)

open import Relation.Binary.PropositionalEquality           using (_≡_; refl)
open import Relation.Binary.HeterogeneousEquality           using (_≅_; refl)

module _ {ℓa ℓp} {A : Set ℓa} (P : A → Set ℓp) where

variable
  a a'  : A
  xs ys : List A
  τ τ'  : xs ⊆ ys
  p     : P a

data AllExt : (τ : xs ⊆ ys) → Set ℓp where
  _∷_  : P a → AllExt τ → AllExt (a ∷ʳ τ)

postulate

  works : {es : AllExt τ} {es' : AllExt τ'}
    → _≅_ {A = AllExt _} (p ∷ es)
          {B = AllExt _} (p ∷ es')

  fails : {es : AllExt τ} {es' : AllExt τ'}
    → _≅_ {A = _} (p ∷ es)
          {B = AllExt _} (p ∷ es')

-- Cannot instantiate the metavariable _95 to solution
-- piSort (Set ℓa)
-- (λ τ.xs →
--    piSort (Set ℓa)
--    (λ τ.ys →
--       piSort (Set ℓa)
--       (λ τ →
--          piSort (Set ℓa)
--          (λ τ' →
--             piSort (Set ℓa)
--             (λ p.a →
--                piSort (Set ℓp)
--                (λ p →
--                   piSort (Set ℓp)
--                   (λ es →
--                      piSort (Set ℓp)
--                      (λ es' →
--                         Set
--                         (ℓa Agda.Primitive.⊔ ℓp Agda.Primitive.⊔ _a_132 (P = P))))))))))
-- since it contains the variable ℓa
-- which is not in scope of the metavariable or irrelevant in the metavariable but relevant in the solution
-- when checking that the expression
-- {es : AllExt τ} {es' : AllExt τ'} →
-- τ ≡ τ' → es ≅ es' → _≅_ {A = _} (p ∷ es) {B = AllExt _} (p ∷ es')
-- has type _95
```
I find this error quite incomprehensible, because why should `ℓa`, a global module parameter, not be in scope of a metavariable?  Or why irrelevant?

Note: Error becomes "Unsolved constraints" if we make es and es' module parameters instead.

The problem seems to involve overloaded constructors, since it is triggered by removing the manual disambiguation.

**Comment by jespercockx** [comment]
After the fix of #3982, this now gives unsolved constraints:
```
Sort _87  [ at /home/jesper/Library/Programming/Agda/Issue3940.agda:25,11-27,35 ]
_a_116 : Agda.Primitive.Level  [ at /home/jesper/Library/Programming/Agda/Issue3940.agda:26,7-10 ]
_A_117 : Set (_a_116 (P = P))  [ at /home/jesper/Library/Programming/Agda/Issue3940.agda:26,16-17 ]
_118 : _A_117 (P = P)  [ at /home/jesper/Library/Programming/Agda/Issue3940.agda:26,20-26 ]

———— Errors ————————————————————————————————————————————————
Failed to solve the following constraints:
  univSort
  (piSort (Set ℓa)
   (λ τ.xs →
      piSort (Set ℓa)
      (λ τ.ys →
         piSort (Set ℓa)
         (λ τ →
            piSort (Set ℓa)
            (λ τ'.xs →
               piSort (Set ℓa)
               (λ τ'.ys →
                  piSort (Set ℓa)
                  (λ τ' →
                     piSort (Set ℓa)
                     (λ p.a →
                        piSort (Set ℓp)
                        (λ p →
                           piSort (Set ℓp)
                           (λ es → piSort (Set ℓp) (λ es' → Set (_a_116 (P = P)))))))))))))
    = univSort _87
  piSort (Set ℓa)
  (λ τ.xs →
     piSort (Set ℓa)
     (λ τ.ys →
        piSort (Set ℓa)
        (λ τ →
           piSort (Set ℓa)
           (λ τ'.xs →
              piSort (Set ℓa)
              (λ τ'.ys →
                 piSort (Set ℓa)
                 (λ τ' →
                    piSort (Set ℓa)
                    (λ p.a →
                       piSort (Set ℓp)
                       (λ p →
                          piSort (Set ℓp)
                          (λ es → piSort (Set ℓp) (λ es' → Set (_a_116 (P = P))))))))))))
    = _87
```

Is this the expected behaviour?

**Comment by UlfNorell** [comment]
Closing as fixed.
